promql/engine.go

```diff
@@ -3758,57 +3749,65 @@ func preprocessExprHelper(expr parser.Expr, start, end time.Time) bool {
 		case parser.END:
 			n.Timestamp = makeInt64Pointer(timestamp.FromTime(end))
 		}
-		return n.Timestamp != nil
+		return n.Timestamp != nil, n.Timestamp != nil
 
 	case *parser.AggregateExpr:
+		unwrapParenExpr(&n.Expr)
+		unwrapParenExpr(&n.Param)
 		return preprocessExprHelper(n.Expr, start, end)
 
 	case *parser.BinaryExpr:
-		isInvariant1, isInvariant2 := preprocessExprHelper(n.LHS, start, end), preprocessExprHelper(n.RHS, start, end)
+		isInvariant1, shouldWrap1 := preprocessExprHelper(n.LHS, start, end)
+		isInvariant2, shouldWrap2 := preprocessExprHelper(n.RHS, start, end)
 		if isInvariant1 && isInvariant2 {
-			return true
+			return true, true
 		}
 
-		if isInvariant1 {
+		if shouldWrap1 {
 			n.LHS = newStepInvariantExpr(n.LHS)
 		}
-		if isInvariant2 {
+		if shouldWrap2 {
 			n.RHS = newStepInvariantExpr(n.RHS)
 		}
 
-		return false
+		return false, false
 
 	case *parser.Call:
 		_, ok := AtModifierUnsafeFunctions[n.Func.Name]
 		isStepInvariant := !ok
-		isStepInvariantSlice := make([]bool, len(n.Args))
+		shouldWrap := make([]bool, len(n.Args))
 		for i := range n.Args {
-			isStepInvariantSlice[i] = preprocessExprHelper(n.Args[i], start, end)
-			isStepInvariant = isStepInvariant && isStepInvariantSlice[i]
+			unwrapParenExpr(&n.Args[i])
+			var argIsStepInvariant bool
+			argIsStepInvariant, shouldWrap[i] = preprocessExprHelper(n.Args[i], start, end)
+			isStepInvariant = isStepInvariant && argIsStepInvariant
 		}
 
 		if isStepInvariant {
 			// The function and all arguments are step invariant.
-			return true
+			return true, true
 		}
 
-		for i, isi := range isStepInvariantSlice {
+		for i, isi := range shouldWrap {
 			if isi {
 				n.Args[i] = newStepInvariantExpr(n.Args[i])
 			}
 		}
-		return false
+		return false, false
 
 	case *parser.MatrixSelector:
-		return preprocessExprHelper(n.VectorSelector, start, end)
+		// We don't need to wrap a MatrixSelector because functions over range vectors evaluate those directly,
+		// and they can't appear at top level in a range query.
```

Comment:
On second thoughts, I think this second line of comment is unnecessary.
It was provoked by a test failing.
So long as all evaluations of MatrixSelector check the Timestamp field, we don't need a wrapper.

Reply:
i think this would change after https://github.com/prometheus/prometheus/pull/14095? post that PR, the `MatrixSelector` can appear in a range query and it will be evaluated in an `eval` call, so we'd want it to be wrapped in `StepInvariantExpr` if it is indeed step invariant
eg:

```
rate(my_metric[5m] @ start() != 0)
```